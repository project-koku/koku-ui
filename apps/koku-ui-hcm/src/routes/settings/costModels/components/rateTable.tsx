import { Tooltip } from '@patternfly/react-core';
import type { IActions, ThProps } from '@patternfly/react-table';
import {
  ActionsColumn,
  ExpandableRowContent,
  Table,
  TableVariant,
  Tbody,
  Td,
  Th,
  Thead,
  Tr,
} from '@patternfly/react-table';
import type { Rate } from 'api/rates';
import { intl as defaultIntl } from 'components/i18n';
import messages from 'locales/messages';
import React from 'react';
import type { WrappedComponentProps } from 'react-intl';
import { injectIntl } from 'react-intl';
import { formatCurrencyRate, unitsLookupKey } from 'utils/format';

interface RateTableProps extends WrappedComponentProps {
  actions?: IActions;
  tiers: Rate[];
  sortCallback?: ({ index, direction }) => void;
}

// defaultIntl required for testing
const RateTableBase: React.FC<RateTableProps> = ({
  actions = [],
  intl = defaultIntl,
  tiers,
  sortCallback = () => {},
}) => {
  const [activeSortIndex, setActiveSortIndex] = React.useState<number | null>(null);
  const [activeSortDirection, setActiveSortDirection] = React.useState<'asc' | 'desc' | null>(null);
  const [expanded, setExpanded] = React.useState([]);

  const getMetric = value => intl.formatMessage(messages.metricValues, { value }) || value;
  const getMeasurement = (measurement, units) => {
    units = intl.formatMessage(messages.units, { units: unitsLookupKey(units) }) || units;
    return intl.formatMessage(messages.measurementValues, {
      value: measurement.toLowerCase().replace('-', '_'),
      units,
      count: 2,
    });
  };

  const rows = tiers.reduce((acc, tier, ix) => {
    const isTagRates = !tier.tiered_rates;
    const tierRate = tier.tiered_rates ? tier.tiered_rates[0].value : 0;
    return [
      ...acc,
      {
        data: { index: ix, hasChildren: isTagRates, tag_rates: tier.tag_rates, stateIndex: tier.stateIndex },
        cells: [
          getMetric(tier.metric.label_metric),
          tier.description || '',
          getMeasurement(tier.metric.label_measurement, tier.metric.label_measurement_unit),
          tier.cost_type,
          {
            title: isTagRates
              ? intl.formatMessage(messages.various)
              : formatCurrencyRate(tierRate, tier.tiered_rates[0].unit),
            expandToggle: isTagRates,
          },
        ],
      },
    ];
  }, []);
  const columns = [
    { title: intl.formatMessage(messages.metric), sortable: true, sortIndex: 0 },
    { title: intl.formatMessage(messages.description) },
    { title: intl.formatMessage(messages.measurement), sortable: true, sortIndex: 2 },
    { title: intl.formatMessage(messages.calculationType) },
    { title: intl.formatMessage(messages.rate) },
  ];
  const tagColumns = [
    intl.formatMessage(messages.costModelsTagRateTableKey),
    intl.formatMessage(messages.costModelsTagRateTableValue),
    intl.formatMessage(messages.rate),
    intl.formatMessage(messages.description),
    intl.formatMessage(messages.default),
  ];

  const getSortParams = (columnIndex: number): ThProps['sort'] => ({
    sortBy: {
      index: activeSortIndex,
      direction: activeSortDirection,
      defaultDirection: 'asc',
    },
    onSort: (_evt, index, direction) => {
      setActiveSortIndex(index);
      setActiveSortDirection(direction);
      sortCallback({ index, direction });
    },
    columnIndex,
  });
  const setRowExpanded = rowIndex => {
    setExpanded(expanded.includes(rowIndex) ? expanded.filter(ex => ex !== rowIndex) : expanded.concat([rowIndex]));
  };
  const compoundExpandParams = rowIndex => ({
    isExpanded: expanded.includes(rowIndex),
    onToggle: () => setRowExpanded(rowIndex),
    expandId: 'expand-' + rowIndex,
    rowIndex,
    columnIndex: 4,
  });

  const sortedRows =
    activeSortIndex === null
      ? rows
      : rows.sort((a, b) => {
          const aValue = a.cells[activeSortIndex];
          const bValue = b.cells[activeSortIndex];
          if (activeSortDirection === 'asc') {
            return (aValue as string).localeCompare(bValue as string);
          }
          return (bValue as string).localeCompare(aValue as string);
        });

  return (
    <Table
      aria-label={intl.formatMessage(messages.costModelsWizardCreatePriceList)}
      hasAnimations
      isExpandable
      variant={TableVariant.compact}
    >
      <Thead>
        <Tr>
          {columns.map((col: { title?: string; sortable?: boolean }, i) => (
            <Th key={i} sort={col.sortable ? getSortParams(i) : undefined}>
              {col.title}
            </Th>
          ))}
          {!!actions.length && <Th></Th>}
        </Tr>
      </Thead>
      {sortedRows.map((row, rowIndex) => {
        const rowId = `row-${rowIndex}`;
        const isExpanded = row.data.hasChildren && expanded.includes(rowIndex);
        return (
          <Tbody isExpanded={isExpanded} key={rowId}>
            <Tr isContentExpanded={isExpanded} isControlRow key={`${rowId}-${rowIndex}`}>
              {row.cells.map((cell, i) => (
                <Td
                  key={i}
                  compoundExpand={
                    cell.expandToggle && row.data.hasChildren ? compoundExpandParams(rowIndex) : undefined
                  }
                >
                  {cell.title ? cell.title : cell}
                </Td>
              ))}
              {!!actions.length && (
                <Td key={row.cells.length} isActionCell>
                  <Tooltip content={intl.formatMessage(messages.moreOptions)}>
                    <ActionsColumn
                      items={actions.map(a => {
                        return {
                          ...a,
                          onClick: () => {
                            a.onClick(null, rowIndex, row, null);
                          },
                        };
                      })}
                    />
                  </Tooltip>
                </Td>
              )}
            </Tr>
            {row.data.hasChildren && (
              <Tr isExpanded={isExpanded}>
                <Td colSpan={6}>
                  <ExpandableRowContent>
                    <Table borders={false} variant={TableVariant.compact}>
                      <Thead>
                        <Tr>
                          {tagColumns.map((tag, tagIndex) => (
                            <Th key={tagIndex}>{tag}</Th>
                          ))}
                        </Tr>
                      </Thead>
                      <Tbody>
                        {row.data.tag_rates.tag_values.map((v, index) => (
                          <Tr key={index}>
                            <Td>{index === 0 ? row.data.tag_rates.tag_key : ''}</Td>
                            <Td>{v.tag_value}</Td>
                            <Td>{formatCurrencyRate(v.value, v.unit)}</Td>
                            <Td>{v.description}</Td>
                            <Td>{v.default ? intl.formatMessage(messages.yes) : intl.formatMessage(messages.no)}</Td>
                          </Tr>
                        ))}
                      </Tbody>
                    </Table>
                  </ExpandableRowContent>
                </Td>
              </Tr>
            )}
          </Tbody>
        );
      })}
    </Table>
  );
};

const RateTable = injectIntl(RateTableBase);
export { RateTable };
