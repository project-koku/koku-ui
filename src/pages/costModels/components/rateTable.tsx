import {
  compoundExpand,
  IActions,
  ISortBy,
  sortable,
  SortByDirection,
  Table,
  TableBody,
  TableHeader,
  TableVariant,
} from '@patternfly/react-table';
import { Rate } from 'api/rates';
import React from 'react';
import { formatCurrency } from 'utils/formatValue';

import { compareBy } from './rateForm/utils';
import TagRateTable from './tagRateTable';

interface RateTableProps {
  tiers: Rate[];
  actions?: IActions;
  isCompact?: boolean;
  t?: any;
}

export const RateTable: React.SFC<RateTableProps> = ({ t = v => v, tiers, actions, isCompact }) => {
  const [expanded, setExpanded] = React.useState({});
  const [sortBy, setSortBy] = React.useState<ISortBy>({});
  const cells = [
    { title: t('description') },
    { title: t('cost_models.table.metric'), transforms: [sortable] },
    { title: t('cost_models.table.measurement'), transforms: [sortable] },
    { title: t('cost_models.calculation_type') },
    { title: t('cost_models.rate'), cellTransforms: [compoundExpand] },
  ];
  const onSort = (_event, index: number, direction: SortByDirection) => {
    setSortBy({ index, direction });
  };
  let counter = 0;
  const rows = tiers
    .sort((r1, r2) => {
      const projection =
        sortBy.index === 1
          ? (r: Rate) => r.metric.label_metric
          : sortBy.index === 2
          ? (r: Rate) => r.metric.label_measurement
          : () => '';
      return compareBy(r1, r2, sortBy.direction, projection);
    })
    .reduce((acc, tier, ix) => {
      const rateKind = tier.tiered_rates ? 'regular' : 'tagging';
      let compoundRows = [];
      if (rateKind === 'tagging') {
        compoundRows = [
          {
            compoundParent: 4,
            parent: ix + counter,
            cells: [
              {
                title: <TagRateTable tagRates={tier.tag_rates} />,
                props: { colSpan: 6, className: 'pf-m-no-padding' },
              },
            ],
          },
        ];
        counter += 1;
      }
      const isOpen = rateKind === 'tagging' ? expanded[ix + counter - 1] || false : undefined;
      return [
        ...acc,
        {
          data: { index: ix, hasChildren: rateKind === 'tagging' },
          cells: [
            tier.description || '',
            tier.metric.label_metric,
            `${tier.metric.label_measurement} (${tier.metric.label_measurement_unit})`,
            tier.cost_type,
            {
              title:
                rateKind === 'regular'
                  ? `${formatCurrency(Number(tier.tiered_rates[0].value), 'USD')}`
                  : t('cost_models.table.tagged_rates'),
              props: { isOpen, style: { padding: rateKind === 'tagging' ? '' : '1.5rem 1rem' } },
            },
          ],
        },
        ...compoundRows,
      ];
    }, []);
  const onExpand = (_event: any, rowIndex: number, _colIndex: number, isOpen: boolean) => {
    setExpanded({ ...expanded, [rowIndex]: !isOpen });
  };
  const actionResolver = rowData => {
    if (rowData.compoundParent) {
      return [];
    }
    return actions;
  };
  return (
    <Table
      onSort={onSort}
      sortBy={sortBy}
      aria-label="price list"
      variant={isCompact ? TableVariant.compact : undefined}
      rows={rows}
      cells={cells}
      actionResolver={actionResolver}
      onExpand={onExpand}
    >
      <TableHeader />
      <TableBody />
    </Table>
  );
};
