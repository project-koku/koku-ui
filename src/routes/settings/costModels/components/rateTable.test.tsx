import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import type { Rate } from 'api/rates';
import React from 'react';

import { RateTable } from './rateTable';

describe('rate-table', () => {
  test('smoke-test', async () => {
    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
    const tiers: Rate[] = [
      {
        description: 'rate 1',
        metric: {
          name: 'cpu_core_request_per_hour',
          metric: 'cpu_core_request_per_hour',
          source_type: 'openshift container platform',
          default_cost_type: 'Supplementary',
          label_metric: 'CPU',
          label_measurement: 'Request',
          label_measurement_unit: 'core-hour',
        },
        cost_type: 'Infrastructure',
        tiered_rates: [
          {
            unit: 'USD',
            value: 130.32,
            usage: {
              unit: 'USD',
            },
          },
        ],
      },
      {
        description: 'rate 2',
        metric: {
          name: 'cpu_core_request_per_hour',
          metric: 'cpu_core_request_per_hour',
          source_type: 'openshift container platform',
          label_metric: 'CPU',
          label_measurement: 'Usage',
          label_measurement_unit: 'core-hour',
          default_cost_type: 'Supplementary',
        },
        cost_type: 'Supplementary',
        tag_rates: {
          tag_key: 'openshift',
          tag_values: [
            {
              unit: 'USD',
              value: 0.43,
              tag_value: 'worker',
              description: 'default',
              default: true,
            },
            {
              unit: 'USD',
              value: 1.5,
              tag_value: 'grafana',
              description: 'grafana containers',
              default: false,
            },
          ],
        },
      },
    ];
    render(<RateTable tiers={tiers} />);
    expect(screen.getByText('rate 1')).toBeTruthy();
    expect(screen.getByText('rate 2')).toBeTruthy();
    expect(screen.queryByText('grafana')).toBeNull();
    await act(async () => user.click(screen.getByRole('button', { name: 'Various' })));
    expect(screen.getByText('grafana')).toBeTruthy();
  });
  test('sort by metric & measurement', () => {
    // eslint-disable-next-line no-console
    console.error = jest.fn();
    const tiers: Rate[] = [
      {
        description: '',
        metric: {
          name: 'node_cost_per_month',
          metric: 'node_cost_per_month',
          source_type: 'openshift container platform',
          label_metric: 'Node',
          label_measurement: 'Currency',
          label_measurement_unit: 'node-month',
          default_cost_type: 'Supplementary',
        },
        cost_type: 'Supplementary',
        tiered_rates: [
          {
            unit: 'USD',
            value: 125.12,
            usage: {
              unit: 'USD',
            },
          },
        ],
      },
      {
        description: '',
        metric: {
          name: 'cpu_core_request_per_hour',
          metric: 'cpu_core_request_per_hour',
          source_type: 'openshift container platform',
          default_cost_type: 'Supplementary',
          label_metric: 'CPU',
          label_measurement: 'Request',
          label_measurement_unit: 'core-hour',
        },
        cost_type: 'Infrastructure',
        tiered_rates: [
          {
            unit: 'USD',
            value: 5.5,
            usage: {
              unit: 'USD',
            },
          },
        ],
      },
      {
        description: '',
        metric: {
          name: 'cpu_core_request_per_hour',
          metric: 'cpu_core_request_per_hour',
          source_type: 'openshift container platform',
          default_cost_type: 'Supplementary',
          label_metric: 'CPU',
          label_measurement: 'Usage',
          label_measurement_unit: 'core-hour',
        },
        cost_type: 'Infrastructure',
        tiered_rates: [
          {
            unit: 'USD',
            value: 7.2,
            usage: {
              unit: 'USD',
            },
          },
        ],
      },
      {
        description: '',
        metric: {
          name: 'cpu_core_request_per_hour',
          metric: 'cpu_core_request_per_hour',
          source_type: 'openshift container platform',
          default_cost_type: 'Supplementary',
          label_metric: 'CPU',
          label_measurement: 'Request',
          label_measurement_unit: 'core-hour',
        },
        cost_type: 'Supplementary',
        tiered_rates: [
          {
            unit: 'USD',
            value: 124.6,
            usage: {
              unit: 'USD',
            },
          },
        ],
      },
    ];
    render(<RateTable tiers={tiers} />);
    const metrics = screen.getAllByRole('cell', { name: /"value":"(CPU|Node)"/ });
    expect(metrics).toMatchSnapshot();
    userEvent.click(screen.getByRole('button', { name: /metric/i })).then(() => {
      expect(metrics).toMatchSnapshot();
    });
    userEvent.click(screen.getByRole('button', { name: /measurement/i })).then(() => {
      expect(metrics).toMatchSnapshot();
    });
  });
});
