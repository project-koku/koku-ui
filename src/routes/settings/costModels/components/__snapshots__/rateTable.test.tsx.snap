// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rate-table sort by metric & measurement 1`] = `
[
  <td
    class="pf-v5-c-table__td"
    tabindex="-1"
  >
    {value, select, cpu {CPU} cluster {Cluster} memory {Memory} node {Node} persistent_volume_claims {Persistent volume claims} storage {Storage} virtual_machine {Virtual machine}other {}}{"value":"Node"}
  </td>,
  <td
    class="pf-v5-c-table__td"
    tabindex="-1"
  >
    {value, select, cpu {CPU} cluster {Cluster} memory {Memory} node {Node} persistent_volume_claims {Persistent volume claims} storage {Storage} virtual_machine {Virtual machine}other {}}{"value":"CPU"}
  </td>,
  <td
    class="pf-v5-c-table__td"
    tabindex="-1"
  >
    {value, select, cpu {CPU} cluster {Cluster} memory {Memory} node {Node} persistent_volume_claims {Persistent volume claims} storage {Storage} virtual_machine {Virtual machine}other {}}{"value":"CPU"}
  </td>,
  <td
    class="pf-v5-c-table__td"
    tabindex="-1"
  >
    {value, select, cpu {CPU} cluster {Cluster} memory {Memory} node {Node} persistent_volume_claims {Persistent volume claims} storage {Storage} virtual_machine {Virtual machine}other {}}{"value":"CPU"}
  </td>,
]
`;
